Kernel/Net/NetworkTask.cpp

```diff
@@ -329,6 +330,38 @@ void flush_delayed_tcp_acks()
     }
 }
 
+void send_tcp_rst(const IPv4Packet& ipv4_packet, const TCPPacket& tcp_packet, RefPtr<NetworkAdapter> adapter)
+{
+    auto routing_decision = route_to(ipv4_packet.source(), ipv4_packet.destination(), adapter);
+    if (routing_decision.is_zero())
+        return;
+
+    auto ipv4_payload_offset = routing_decision.adapter->ipv4_payload_offset();
+
+    const size_t options_size = 0;
+    const size_t tcp_header_size = sizeof(TCPPacket) + options_size;
+    const size_t buffer_size = ipv4_payload_offset + tcp_header_size;
+
+    auto packet = routing_decision.adapter->acquire_packet_buffer(buffer_size);
+    if (!packet)
+        return;
+    routing_decision.adapter->fill_in_ipv4_header(*packet, ipv4_packet.destination(),
+        routing_decision.next_hop, ipv4_packet.source(), IPv4Protocol::TCP,
+        buffer_size - ipv4_payload_offset, 64);
+    memset(packet->buffer.data() + ipv4_payload_offset, 0, sizeof(TCPPacket));
+    auto& rst_packet = *(TCPPacket*)(packet->buffer.data() + ipv4_payload_offset);
+    rst_packet.set_source_port(tcp_packet.destination_port());
+    rst_packet.set_destination_port(tcp_packet.source_port());
+    rst_packet.set_window_size(0);
+    rst_packet.set_sequence_number(0);
+    rst_packet.set_ack_number(tcp_packet.sequence_number() + 1);
+    rst_packet.set_data_offset(tcp_header_size / sizeof(u32));
+    rst_packet.set_flags(TCPFlags::RST | TCPFlags::ACK);
+
+    routing_decision.adapter->send_packet({ packet->buffer.data(), packet->buffer.size() });
```

Comment:
You need to calculate the TCP checksum before sending the packet (`rst_packet.set_checksum()`), otherwise this packet will get discarded by other hosts.

Reply:
![image](https://user-images.githubusercontent.com/388571/127776543-5ff3395f-1102-4b42-ab60-c9fdd6576e80.png)

Reply:
Looks better now. Required making the compute_tcp_checksum public.

Reply:
![wireshark](https://user-images.githubusercontent.com/54456386/127779494-3b170dfc-51c4-4427-b6ad-f4014ac7b0b6.png)